src/logic/map_objects/immovable_program.cc

```diff
@@ -110,94 +201,152 @@ void ImmovableProgram::ActPlaySound::execute(Game& game, Immovable& immovable) c
-	if (immovable.apply_growth_delay(game)) {
-		return;
-	}
-	if (probability == 0 || game.logic_rand() % 256 < probability) {
-		Player* player = immovable.get_owner();
-		Coords const c = immovable.get_position();
-		MapObjectDescr::OwnerType owner_type = immovable.descr().owner_type();
-		immovable.remove(game);  //  Now immovable is a dangling reference!
-
-		if (bob) {
-			game.create_ship(c, type_name, player);
-		} else {
-			game.create_immovable_with_name(
-			   c, type_name, owner_type, player, nullptr /* former_building_descr */);
-		}
-	} else {
-		immovable.program_step(game);
-	}
-}
+	assert(removal_wanted || !type_name.empty());
 
-ImmovableProgram::ActGrow::ActGrow(std::vector<std::string>& arguments,
-                                   const ImmovableDescr& descr) {
-	if (arguments.size() != 1) {
-		throw GameDataError("Usage: grow=<immovable name>");
+	// Check whether we want to remove
+	if (removal_wanted &&
+	    (removal_probability == 0 || game.logic_rand() % 256 < removal_probability)) {
+		immovable.remove(game);  //  Now immovable is a dangling reference!
+		return;
 	}
-	if (!descr.has_terrain_affinity()) {
-		throw GameDataError(
-		   "Immovable %s can 'grow', but has no terrain_affinity entry.", descr.name().c_str());
+	// For removal with chance but no transformation to new object
+	if (type_name.empty()) {
+		return;
 	}
 
-	// TODO(GunChleoc): If would be nice to check if target exists, but we can't guarantee the load
-	// order. Maybe in postload() one day.
-	type_name = arguments.front();
-}
-
-void ImmovableProgram::ActGrow::execute(Game& game, Immovable& immovable) const {
 	if (immovable.apply_growth_delay(game)) {
 		return;
 	}
 
+	// Transform according to success if this was specified, and using any available terrain affinity
+	// if not.
 	const Map& map = game.map();
-	FCoords const f = map.get_fcoords(immovable.get_position());
 	const ImmovableDescr& descr = immovable.descr();
+	Player* player = immovable.get_owner();
+	FCoords const f = map.get_fcoords(immovable.get_position());
+	MapObjectDescr::OwnerType owner_type = immovable.descr().owner_type();
 
-	if ((game.logic_rand() % TerrainAffinity::kPrecisionFactor) <
-	    probability_to_grow(descr.terrain_affinity(), f, map, game.world().terrains())) {
-		MapObjectDescr::OwnerType owner_type = descr.owner_type();
-		Player* owner = immovable.get_owner();
-		immovable.remove(game);  //  Now immovable is a dangling reference!
-		game.create_immovable_with_name(
-		   f, type_name, owner_type, owner, nullptr /* former_building_descr */);
+	const bool will_transform =
+	   descr.has_terrain_affinity() && transform_probability == 0 ?
+	      (game.logic_rand() % TerrainAffinity::kPrecisionFactor) <
+	         probability_to_grow(descr.terrain_affinity(), f, map, game.world().terrains()) :
+	      transform_probability == 0 || game.logic_rand() % 256 < transform_probability;
+
+	if (will_transform) {
+		if (bob) {
+			game.create_ship(f, type_name, player);
```

Comment:
No need to delete the immovable afterwards?

Reply:
Thanks for catching this bug